### PR TITLE
Add Methods and Tests to Document Model

### LINKED ISSUE #3: Create the Document model
# Overview
Much of the functionality of the [`gender_analysis`](https://github.com/dhmit/gender_analysis) project centers around a class called [`Document`](https://github.com/dhmit/gender_analysis/blob/main/gender_analysis/text/document.py). Instances of this class are created with reference to a `.txt` file on the user's machine, and the instance provides an interface for engaging with the textual contents of that file. Properties and methods of this class include [`word_count`](https://github.com/dhmit/gender_analysis/blob/main/gender_analysis/text/document.py#L87), [`_clean_text`](https://github.com/dhmit/gender_analysis/blob/main/gender_analysis/text/document.py#L240), and [`get_tokenized_text`](https://github.com/dhmit/gender_analysis/blob/main/gender_analysis/text/document.py#L310).

This basic concept, that the `Document` class provides an interface for engaging with textual material, will be a useful tool as we move the whole project into a web application architecture. That said, we should replace some of the conventions of this `Document` class with more suitable web equivalents.

## Goals
- Port over the `Document` class into `gender_analysis_web` as a Django model.
- Determine what we should persist in the database record that underpins this new `Document` model. For instance: the existing `Document` class memoizes the results of `get_tokenized_text`, whereas we can (and probably should) store those results in the database record.
- Replace the custom tokenizing implementation in the existing `get_tokenized_text` with an implementation that uses the `nltk` package (for some context, see [this issue](https://github.com/dhmit/gender_analysis/issues/135)).
- Ultimately, allow a user to create new `Document` instances by inputting text content and text metadata into a web interface. To do this, we'll need to create both a user interface and API backend to support this interaction.

### PR DESCRIPTION
**Pull request by phuang00**
This PR transfers over the necessary methods from the `gender_analysis` project for the `Document` model. Method docstrings are updated to reflect changes to the methods itself and tests are written for each method. `Document` field and method names are renamed to follow the same naming convention

### CODE REVIEW

File: backend/app/models.py [hunk 1]
```diff
@@ -44,5 +58,285 @@ class Document(models.Model):
...
+        smart_quotes = {
+            '“': '"',
+            '”': '"',
+            "‘": "'",
+            "’": "'",
+        }
+
+        output_text = self.text
+        for quote in smart_quotes:
+            output_text = output_text.replace(quote, smart_quotes[quote])
```
Comment (MBJean) [inline backend/app/models.py]: I'm wondering about whether or not this is the ideal way to implement this quotation mark clean up. I think that lines 83-84 here will perform four passes on the input text, one for each key in `smart_quotes`, unless `.replace()` is doing something particularly smart that I don't know about. It seems like using the regex `.sub` method would implement this in one pass, and therefore more efficiently. Something like:

```
import re 

re.sub(r'[\“\”\‘\’]', '\"', self.text)
```

File: backend/app/models.py [hunk 2]
```diff
@@ -44,5 +58,285 @@ class Document(models.Model):
...
+        Returns the part of speech tags as a list of tuples. The first part of each tuple is the
+        term, the second one the part of speech tag.
+
+        Note: the same word can have a different part of speech tags.
+
+        :return: List of tuples (term, speech_tag)
+        """
+
+        if not self.part_of_speech_tags:
+            pos_tags = nltk.pos_tag(nltk.word_tokenize(self.text))
```
Comment (MBJean) [inline backend/app/models.py]: I think this `nltk.word_tokenize(self.text)` is performing the tokenization we should assume occurs in `get_tokenized_text`. Can we therefore instead reference just reference `self.tokenized_text`? Although if we follow the pattern I suggest in my comment below, this method might be unnecessary.

File: backend/app/models.py [hunk 3]
```diff
@@ -32,6 +36,16 @@ def __eq__(self, other):
...
 
 
+class DocumentManager(models.Manager):
+    def create_document(self, attributes):
+        doc = self.create(attributes)
+        doc.get_tokenized_text()
+        doc.get_word_count()
+        doc.get_word_count_counter()
+        doc.get_part_of_speech_tags()
+        return doc
```
Comment (MBJean) [inline backend/app/models.py:44]: I like the direction this is going, but I see below that there's still a ton of the vestigial memoization stuff in place, which means there's a ton of boilerplate below that I'm not sure is really adding anything. Could we delete the methods `get_word_count`, `get_word_count_counter`, and `get_part_of_speech_tags` and instead do something like:
```
def create_document(self, attributes):
  text = attributes.text
  tokenized_text = Document.get_tokenized_text(text)
  word_count = len(tokenized_text)
  word_count_counter = Counter(tokenized_text)
  part_of_speech_tags = nltk.pos_tag(tokenized_text)
  document = self.create(#the above kwargs)
  return document
```

Doing so would ensure that the first time a new `Document` instance is saved to the database, all of the relevant fields are populated accordingly, and we can remove any code in the `Document` model that guards against missing fields.

File: backend/app/models.py [hunk 4]
```diff
@@ -44,5 +58,285 @@ class Document(models.Model):
...
+        """
+        Returns the number of instances of a word in the text.
+
+        Note: This method is not case sensitive
+
+        :param word: word to be counted in text
+        :return: Number of occurrences of the word, as an int
+        """
+        self.get_word_count_counter()
+        return self.word_count_counter[word.lower()]
```
Comment (MBJean) [inline backend/app/models.py]: Per my comment above about streamlining these methods, if we followed that plan, this method becomes:
```
return self.word_count_counter[word.lower()]
```

Or, to guard against the non-existence of the word in the counter, maybe something like:
```
try:
  return self.word_count_counter[word.lower()]
except KeyError:
  return 0
```

**Review by MBJean — CHANGES_REQUESTED** [review]
I think you've done a great job bringing over the old work and adapting it to the new context. Great work! I recommend we take this a bit further: many of the methods in the current `Document` model exist for the purpose of memoizing expensive operations, something that we don't necessarily need to do any longer. I think we can further front load the work in the `create_document` method and then go back and delete a bunch of methods and clean others up. Check out my comments above and let me know what you think.

File: backend/app/models.py (hunk 3 above)
Comment (phuang00) [reply backend/app/models.py:44]: I've moved the functionalities of `get_word_count`, `get_word_count_counter`, and `get_part_of_speech_tags` into `get_tokenized_text` and renamed it `get_tokenized_text_wc_and_pos`. This way, we'll only need to call `get_tokenized_text_wc_and_pos` when we first add in the `text` and when we update it.

File: backend/app/models.py (hunk 4 above)
Comment (phuang00) [reply backend/app/models.py]: fixed!

**Comment by phuang00** [comment]
I've implemented the changes above and it declutters a lot of the code. We didn't touch a lot of the memoization stuff before because we feared that the methods would be called in the wrong orders, but with the frontend interface and `create_document`, this won't be a problem anymore.

**Comment by phuang00** [comment]
`views.py` has been updated to use `create_document` to create instances of `Document` and the naming has been switched from `add_text` to `add_document`.

**Review by MBJean — APPROVED** [review]
👍  This LGTM! Thanks for all your hard work on this. Merging.

**Pull request merged**